pkg/controller/apiservicecabundle/controller/apiservice_cabundle_controller.go

```diff
@@ -19,89 +13,45 @@ import (
 	"github.com/openshift/service-serving-cert-signer/pkg/controller/api"
 )
 
-type ServiceServingCertUpdateController struct {
+type serviceServingCertUpdateController struct {
 	apiServiceClient apiserviceclient.APIServicesGetter
 	apiServiceLister apiservicelister.APIServiceLister
 
 	caBundle []byte
-
-	// services that need to be checked
-	queue workqueue.RateLimitingInterface
-
-	// standard controller loop
-	*controller.Controller
 }
 
-func NewAPIServiceCABundleInjector(apiServiceInformer apiserviceinformer.APIServiceInformer, apiServiceClient apiserviceclient.APIServicesGetter, caBundle []byte) *ServiceServingCertUpdateController {
-	sc := &ServiceServingCertUpdateController{
+func NewAPIServiceCABundleInjector(apiServiceInformer apiserviceinformer.APIServiceInformer, apiServiceClient apiserviceclient.APIServicesGetter, caBundle []byte) controller.Runner {
+	sc := &serviceServingCertUpdateController{
 		apiServiceClient: apiServiceClient,
 		apiServiceLister: apiServiceInformer.Lister(),
 		caBundle:         caBundle,
 	}
 
-	apiServiceInformer.Informer().AddEventHandler(
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    sc.addAPIService,
-			UpdateFunc: sc.updateAPIService,
-		},
-	)
-
-	internalController, queue := controller.New("APIServiceCABundleInjector", sc.syncAPIService, apiServiceInformer.Informer().GetController().HasSynced)
-
-	sc.Controller = internalController
-	sc.queue = queue
+	return controller.New("APIServiceCABundleInjector", sc).
+		WithInformer(apiServiceInformer.Informer(), controller.FilterFuncs{
+			AddFunc:    api.HasInjectCABundleAnnotation,
+			UpdateFunc: api.HasInjectCABundleAnnotationUpdate,
+		})
+}
 
-	return sc
+func (c *serviceServingCertUpdateController) Key(namespace, name string) (v1.Object, error) {
+	return c.apiServiceLister.Get(name)
 }
 
-func (c *ServiceServingCertUpdateController) syncAPIService(obj interface{}) error {
-	key := obj.(string)
-	_, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		return err
-	}
+func (c *serviceServingCertUpdateController) Sync(obj v1.Object) error {
+	apiService := obj.(*apiregistrationv1.APIService)
```

Comment:
We should probably add a nil check here, just in case. api.HasInjectCABundleAnnotation() does a dereference of the passed-in object.